azkaban-common/src/test/java/azkaban/utils/FileIOUtilsTest.java

```diff
@@ -116,6 +117,7 @@ public void testHardlinkCopy() throws IOException {
     assertThat(areDirsEqual(this.baseDir, this.sourceDir, true)).isTrue();
   }
 
+  @Ignore("TODO too slow: over 30s - should this be deleted entirely?")
```

Comment:
thanks for reporting. I agree it's slow. But the purpose of this test is to verify the use case where hard linking a very long dir won't cause issue which happened before in our production. The old code was using linux shell command to do the hard link which will generate excessively long command exceeding the argument limit size.

Reply:
Do you think that there's a substantial risk that the code would be changed to something else to cause regression?

Any way, I suggest:
- have this test `@Ignore`'d
- add a good comment in the actual code why it's done like it is now (maybe point out that there's an ignored unit test available)

This way it shouldn't be possible that anyone changes the hard linking code to something else, causing regression. But it allows unit tests serve their purpose of being quick to run while making any changes.

Reply:
@chengren311 thanks for the review.

It's a tradeoff. Ideally, we should invest in some mechanism to allow slower tests to run automatically to flag potential problem but without blocking commits. Linkedin internal testing infrastructure allows it. We should take advantage of it. e.g. have these tests run in our internal wrappers. I'd suggest that we research how other OSS projects handle this. @juhoautio I am interested in your suggestions. Before that, @juhoautio's suggestion makes sense to me. 

We should have a JIRA to track the long-term improvement plan for how to handle slower tests and set standards.

@ameyamk how do you propose we track the tasks? internal JIRA or Github JIRAs?

Reply:
> @juhoautio I am interested in your suggestions

https://github.com/junit-team/junit4/wiki/categories

Reply:
Thanks for the suggestion. This looks like something that can help. We will still need a way to automatically select the different category of tests to run. 

I will invite the team to participate in this discussion.

Reply:
Not a big fan of adding "ignored", I don't think anyone would bother running the ignored tests. I'm curious what is the actual logic we're trying to test. From the code and conversation, we're not actually testing to logic in `createDeepHardlink(..)` but creating a scenario to make sure the `createDeepHardlink(..)` method works. I suggest:
1) figure out the part of logic that has the possibility of causing regression and guard only that part, proper mocking would make it much faster.
or
2) move these scenario based tests out and create another test suite of integration test to validate scenarios.

Reply:
@chengren311 do you have any objection if @juhoautio does what he suggested above?

@oliverhu 
Thanks for your comment. 

I suggest moving forward with this change with some minor modifications ( see this and other comments) and file issues to address the longer-term issues. An alternative is to wait until the improvements you suggested are implemented, which is currently unfunded. cc @ameyamk  IMHO, it is a better tradeoff to make this change as long as we don't delay the longer term investment for too long.

Reply:
@HappyRay @juhoautio 's suggestion makes sense to me.

Reply:
Thanks @chengren311 

@juhoautio do you mind making your suggested changes above?